contracts/creator.cairo

```diff
@@ -0,0 +1,364 @@
+@storage_var
+func game_id_from_game_index(
+        game_index : felt
+    ) -> (
+        game_id : felt
+    ):
+end
+
+@storage_var
+func game_index_from_game_id(
+        game_id : felt
+    ) -> (
+        game_index : felt
+    ):
+end
+
+# Index of most recent game.
+@storage_var
+func latest_game_index(
+    ) -> (
+        game_index : felt
+    ):
+end
+
+# Lets you find the latest state of a given game.
+@storage_var
+func latest_game_generation(
+        game_index : felt
+    ) -> (
+        game_generation : felt
+    ):
+end
+
+# Stores the count of credits.
+@storage_var
+func has_credits(
+        owner_id : felt
+    ) -> (
+        credits : felt
+    ):
+end
+
+# Stores how many games a user has created
+@storage_var
+func user_game_count(
+        owner_id : felt
+    ) -> (
+        count : felt
+    ):
+end
+
+##### Events #####
+@event
+func game_created(
+    owner_id: felt,
+    game_index : felt,
+    game_id : felt,
+    user_game_count : felt
+):
+end
+
+@event
+func contribution_made(
+    user_id : felt,
+    game_index : felt,
+    current_game_generation : felt
+):
+end
+
+@event
+func credit_earned(
+    user_id : felt, 
+    balance : felt
+):
+end
+
+@event
+func credit_reduced(
+    user_id : felt,
+    balance : felt
+):
+end
+
+##################
+@constructor
+func constructor{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }():
+    alloc_locals
+    local acorn = 215679573337205118357336120696157045389097155380324579848828889530384
+    local game_index = 1
+    let (game_id) = hash_game(acorn)
+
+    stored_game.write(
+        game_index=game_index,
+        gen=0,
+        value=acorn
+    )
+
+    let (caller) = get_caller_address()
+
+    # Store the zeroth-index game as owned by the caller of spawn().
+    owner_of_game.write(game_index, caller)
+
+    game_id_from_game_index.write(game_index, game_id)
+    game_index_from_game_id.write(game_id, game_index)
+    latest_game_index.write(game_index)
+    latest_game_generation.write(game_index, 0)
+    game_created.emit(
+        owner_id=caller,
+        game_index=game_index,
+        game_id=game_id,
+        user_game_count=1
+    )
+    return ()
+end
+
+##### Public functions #####
+# Sets the initial state of a game.
+@external
+func create{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_state : felt
+    ):
+    alloc_locals
+
+    let (local caller) = get_caller_address()
+    assert_not_zero(caller)
+
+    let (credits) = has_credits.read(caller)
+    assert_le_felt(CREDIT_REQUIREMENT, credits)
+    has_credits.write(caller, credits - CREDIT_REQUIREMENT)
+
+    # No two games are the same. Game_id == game hash.
+    let (local game_id) = hash_game(game_state)
+    let (existing_index) = game_index_from_game_id.read(game_id)
```

Comment:
I'm not sure how we could validate the game (except maybe checking the resulting array lenght), I guess even wrongly encoded game could yield correct game grid (meaning - it would have correct amount of bits, but in wrong places)

Reply:
By validation I mean checking if no unnecessary bits are set that we don't use.